Modules/_asynciomodule.c

```diff
@@ -434,27 +431,25 @@ future_schedule_callbacks(asyncio_state *state, FutureObj *fut)
         return 0;
     }
 
-    len = PyList_GET_SIZE(fut->fut_callbacks);
-    if (len == 0) {
-        /* The list of callbacks was empty; clear it and return. */
-        Py_CLEAR(fut->fut_callbacks);
-        return 0;
-    }
-
-    for (i = 0; i < len; i++) {
-        PyObject *cb_tup = PyList_GET_ITEM(fut->fut_callbacks, i);
+    // Beware: An evil call_soon could change fut->fut_callbacks.
+    // The idea is to store the callbacks list before clearing it
+    // on the future object. In particular, external code is not
+    // able to mutate the list during the iteration.
+    PyObject *callbacks = Py_NewRef(fut->fut_callbacks);
```

Comment:
You don't really need to incref here, you can just transfer the ownership of list here. That would avoid the unecessary incref/decref